### PR TITLE
add Multi-instance translate chapter in README and its script example 

### PR DESCRIPTION
**Pull request by rongzha1**
(description of the change)
Add Multi-instance translate in README.md and its script example.
Using Multi-instance translate  will have 10+ times speed up.
Here is some data:
AWS EC2 C5.18xlarge:

| Batch | 1 Instance (sent/sec)| 36 Instance (sent/sec)|
| --- | --- | --- |
|   1 |7.50  | 77.172  |           
|   2 |11.30 | 78.46   |                        
|   8 |16.30 | 183.93  |            
|  16 |19.70 | 194.88  |            
|  32 |21.40 | 200.30  |            
|  64 |22.50 | 199.563 |

#### Pull Request Checklist ##
- [] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
 done
- [] Unit tests pass (`pytest`)
didn't change any code.
- [ ] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
no tests modified
- [ ] System tests pass (`pytest test/system`)
 didn't change any code.     
- [ ] Passed code style checking (`./style-check.sh`)
done, no error in README.md and mlt-trans.sh
- [ ] You have considered writing a test
no, didn't change any code.
- [ ] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
not a incompatible change
- [ ] Updated CHANGELOG.md
seems no need to.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

@pengzhao-intel @huangzhiyuan

### CODE REVIEW

**Comment by fhieber** [comment]
I am not sure how generic and re-usable this script is for users (for example by hardcoding newstest2016).

**Comment by rongzha1** [comment]
OK.  We will change script more easy to use. First divided file by instance number, and each instance will handle 1 part. Then put each translated file together to get one file after each instance translate done. @huangzhiyuan

**Comment by rongzha1** [comment]
we have add option for whether running as benchmark.
If run as benchmark, each instance will copy input file as file to be translate.
If not, the input file will be divided by instance number, and each instance will handle one part.
@huangzhiyuan@fhieber

**Comment by rongzha1** [comment]
Can you help to check this PR @fhieber ?  Thanks very much.

File: mlt-trans.sh [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	benchmark $total_cores $model $filetotrans $batch_size
+	sentences=$(($sentences*$loop))
+else
+	realdata $total_cores $model $batch_size
+fi
+end=$(($(date +%s%N)/1000000000))
+
+output=$2".en"
+find . -name "*.result.en" | sort | xargs cat | head -n $line > $output
+rm mlt-gnmt.log* -f
```
Comment (fhieber) [inline mlt-trans.sh]: What is gnmt?

File: mlt-trans.sh [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+# param 1: total_cores
+# param 2: model
+# param 3: file to translate
+# param 4: batch_size
+function benchmark(){
+    for ((i=0; i<$1; i++))
+	do
+		if [ $i == $[$1-1] ]
+		then
+ 			taskset -c $i-$i python3 -m sockeye.translate -m $2 -i $3 -o mlt-en.result.en --batch-size $4 --output-type benchmark --use-cpu > /dev/null 2>&1
```
Comment (fhieber) [inline mlt-trans.sh]: why hardcode the output files?

File: mlt-trans.sh [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	if [ $rema -ne 0 ];then
+		quot=$(($quot+1))
+		cat $file|head -n $(($quot*$workers-$lines)) >> $file
+	fi
+	split -l $quot $file -d -a 2 mlt-gnmt.log.
+	head -n $lines $file > temp.log
+	mv temp.log $file
+}
+
+# benchmark: inference use dummy data
```
Comment (fhieber) [inline mlt-trans.sh]: what kind of dummy data?

**Comment by rongzha1** [comment]
hi @fhieber
Thanks for your quick response!
We will change sh script to python script.
Agree to remove speed numbers.
Will make README.md more acurrate and clearly.  "translate 24 instances in C4.8xlarge" , means use 24 instances to translate on C4.8xlarge platform. (Actually 24 is typo, we use 18 instances in C4.8xlarge for C4.8xlarge has 18 psychical cores each socket)
Assume script can be run on multi-core cpu environment. We test it on broadwell and skylake servers . It works well.  I am not sure whether GPU can benefit from multi-instance. 

We will add some brief descriptions on multi-instances translation in README.md and move script to tutorial/cpu_benchmarking

File: mlt-trans.sh (hunk 1 above)
Comment (rongzha1) [reply mlt-trans.sh]: Actually , this log is not used.  We will rm it in next version.

File: mlt-trans.sh (hunk 2 above)
Comment (rongzha1) [reply mlt-trans.sh]: OK  will add output file name as  parameters

File: mlt-trans.sh (hunk 3 above)
Comment (rongzha1) [reply mlt-trans.sh]: we use input file as dummy data but not divided. That is to say each instance translate the whole input file. This is used as benchmark test.

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    else:
+        start = time.time()
+        merge_translate(total_cores, args)
+        #wait all process complete
+        process_cnt = cores
+        stop_cnt = 0   
+        #force stop 100s after the fisrt instance complete to jump out of the loop
+        while (process_cnt > 0 and stop_cnt < 1000):
+            process_cnt = int(os.popen("ps -ef | grep 'sockeye.translate' | grep -v grep | wc -l").read().split('\n')[0])
+            time.sleep(0.1)
```
Comment (davvil) [inline tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: 0.1 is probably too short time, waking the main process 10 times per second for a process that will take at least minutes to complete. A better solution would probably use os.wait() or similar funtions.

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            stop_cnt += 1
```
Comment (davvil) [inline tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: Wouldn't the stop_cnt trigger for really large files? Also there is no code to differentiate the two exit conditions.

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        end = time.time()
+        #merge each part into one complete output file
+        os.system("find . -name '*.result.en' | sort | xargs cat | head -n %s > %s" % (str(lines), fileOutput))
+        #rm temp file
+        os.system("rm mlt-nmt* -rf")
```
Comment (davvil) [inline tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: This wouldn't be needed with the use of tempfile, as pointed out above.

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        process_cnt = cores
+        stop_cnt = 0   
+        #force stop 100s after the fisrt instance complete to jump out of the loop
+        while (process_cnt > 0 and stop_cnt < 1000):
+            process_cnt = int(os.popen("ps -ef | grep 'sockeye.translate' | grep -v grep | wc -l").read().split('\n')[0])
+            time.sleep(0.1)
+            stop_cnt += 1
+        end = time.time()
+        #merge each part into one complete output file
+        os.system("find . -name '*.result.en' | sort | xargs cat | head -n %s > %s" % (str(lines), fileOutput))
```
Comment (davvil) [inline tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: You only want the files in the current directory, right? A simple 'ls *.result.en' could replace both 'find [...]' and 'sort' (ls sorts its output). Actually 'cat *.results.en' would also work for a number of size below several thousands, which you are actually assuming given the naming schema.

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+def add_args(parser):
+    parser.add_argument("--module" , '-m', help='the pretrained module', default="wmt_model")
+    parser.add_argument("--input_file", '-i', help='the file input to translate', default='newstest2016.tc.BPE.de')
+    parser.add_argument("--output_file", '-o', help='the file saved the result of translation', default='newstest2016.tc.BPE.en')
+    parser.add_argument("--batch_size", '-bs', help='the batch size of process', default=32)
+    parser.add_argument("--data_type", '-t', help='if uses run benchmark or not', default="benchmark")
+
+
+# benchmark: multi-instances translate using same input file
+def benchmark(cores, args):
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: This is for two different use. One is for benchmark,  each core do the whole file inference. The other is that each core do only a part (file_line // core number lines) inference.
If each core do the whole file inference, performance will be better for larger sentence.
If each core do part time file inference, will use less time .

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+    for i in range(cores): 
+        command = ("taskset -c %s-%s python3 -m sockeye.translate -m %s -i %s -o %s --batch-size %s --output-type benchmark --use-cpu > /dev/null 2>&1 " % (str(i), str(i), model, fileInput, fileOutput, str(batchsize)))
+        if i != cores - 1:
+            command += "&"
+        os.system(command)
+
+# split file to small files
+def split_file(cores, fileInput):
+    lines = len(open(fileInput).readlines())
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: Agree. will change later

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    quot = lines // cores
+    rema = lines % cores
+    if rema != 0:
+        quot += 1
+        os.system("cp %s mlt-nmt-temp.log && head -n%s %s >> mlt-nmt-temp.log " % (fileInput, str(quot * cores - lines), fileInput))
+
+    os.system("split -l %s mlt-nmt-temp.log -d -a 2 mlt-nmt.log." % str(quot))
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: Agree. will change later

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+# merge_translate: multi-instances translation, each instance trans  whole_lines/instance_num lines of file, and merge into one complete output file
+def merge_translate(cores, args):
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: Agree. will change this function later

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    model = args.module
+    batchsize = args.batch_size
+
+    # split inputfile to a series of small files which number is equal cores
+    split_file(cores, fileInput)
+    for i in range(cores):
+        ifile = ''
+        if i < 10:
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: good idea. will change later

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+            ifile = "mlt-nmt.log.0" + str(i) 
+        else:
+            ifile = "mlt-nmt.log." + str(i)
+
+        command = ("taskset -c %s-%s python3 -m sockeye.translate -m %s -i %s -o %s --batch-size %s --output-type benchmark --use-cpu > /dev/null 2>&1 " % (str(i), str(i), model, ifile, ifile+".result.en", str(batchsize)))
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: OK. will change later

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        os.system(command)
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='MXnet Sockeye Multi-instances Benchmark')
+    add_args(parser)
+    args = parser.parse_args()
+    fileInput = args.input_file
+    fileOutput = args.output_file
+
+    socket = int(os.popen("grep -w 'physical id' /proc/cpuinfo | sort -u | wc -l").read().split('\n')[0])
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: OK. will change later

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    cores = int(os.popen("grep -w 'core id' /proc/cpuinfo | sort -u | wc -l").read().split('\n')[0])
+    total_cores = socket * cores
+
+    os.system("export KMP_AFFINITY=granularity=fine,noduplicates,compact,1,0")
+    ompStr = "export OMP_NUM_THREADS=" + str(total_cores)
+    os.system(ompStr)
+
+    # the total lines of input file will be translated
+    lines = len(open(args.input_file).readlines())
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: Yes. will change later

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py (hunk 4 above)
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: OK, I'll change to os.wait

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py (hunk 5 above)
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: OK. will use os.wait instead

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py (hunk 7 above)
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: OK.will change later

File: README.md [hunk 16]
```diff
@@ -161,6 +161,15 @@ You can translate as follows:
 This will take the best set of parameters found during training and then translate strings from STDIN and
 write translations to STDOUT.
 
+#### Multi-instance Translate
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply README.md]: OK

File: README.md [hunk 17]
```diff
@@ -161,6 +161,15 @@ You can translate as follows:
...
+Multi-instance can be used to greatly speedup translation in one multi-core processor computer.
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply README.md]: OK

File: tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py [hunk 18]
```diff
@@ -0,0 +1,129 @@
+# Describtion: This script is used for CPU Multi-instance Translate,
```
[the remark this answers is not in the document]
Comment (rongzha1) [reply tutorials/cpu_benchmarking/mlt_cpu_trans_benchmark.py]: OK

File: tutorials/process_per_core_translation/README.md [hunk 19]
```diff
@@ -0,0 +1,13 @@
+# CPU process per core translation
+On multi-core processor computer, translation per core separately can speedup translation performance, due to some operation can't be handled parallel in one process.
+Using this methord, translation on each core can be parallel.
```
Comment (tdomhan) [inline tutorials/process_per_core_translation/README.md]: typo: method

File: tutorials/process_per_core_translation/README.md [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+One python script example is givne and you can run it as follows:
```
Comment (tdomhan) [inline tutorials/process_per_core_translation/README.md]: typo: given

File: tutorials/process_per_core_translation/cpu_process_per_core_translation.py [hunk 21]
```diff
@@ -0,0 +1,129 @@
+# Describtion: This script is used for CPU Multi-instance Translate, which process per core translation.
+# It can greatly speedup translate perefomance.
+# FileName: cpu_process_per_core_translation.py
```
Comment (tdomhan) [inline tutorials/process_per_core_translation/cpu_process_per_core_translation.py]: filename and version seem unnecessary. The usage would be best as part of the argparse description `argparse.ArgumentParser(description=...`.

File: tutorials/process_per_core_translation/cpu_process_per_core_translation.py [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+    parser.add_argument("--module" , '-m', help='the pretrained module', default="wmt_model")
+    parser.add_argument("--input_file", '-i', help='the file input to translate', default='newstest2016.tc.BPE.de')
+    parser.add_argument("--output_file", '-o', help='the file saved the result of translation', default='newstest2016.tc.BPE.en')
+    parser.add_argument("--batch_size", '-bs', help='the batch size of process', default=32)
+    parser.add_argument("--data_type", '-t', help='if uses run benchmark or not', default="benchmark")
+
+def task(args):
+    os.system(args)
+
+# benchmark: multi-instances translating, each instance trans the same input file separately
```
Comment (tdomhan) [inline tutorials/process_per_core_translation/cpu_process_per_core_translation.py]: could you make this a proper doc-string for the method?
```
def benchmark(...):
   """
    doc goes here
   """
```

Also applies for the other functions below.

**Comment by tdomhan** [comment]
thanks for iterating. You will need to rebase on the current master before merging.

File: tutorials/process_per_core_translation/README.md (hunk 19 above)
Comment (rongzha1) [reply tutorials/process_per_core_translation/README.md]: OK

File: tutorials/process_per_core_translation/README.md (hunk 20 above)
Comment (rongzha1) [reply tutorials/process_per_core_translation/README.md]: changed.  Thanks

File: tutorials/process_per_core_translation/cpu_process_per_core_translation.py (hunk 21 above)
Comment (rongzha1) [reply tutorials/process_per_core_translation/cpu_process_per_core_translation.py]: Done. Thanks

File: tutorials/process_per_core_translation/cpu_process_per_core_translation.py (hunk 22 above)
Comment (rongzha1) [reply tutorials/process_per_core_translation/cpu_process_per_core_translation.py]: Good style. will follow.

**Comment by tdomhan** [comment]
thanks for iterating!

**Comment by davvil** [comment]
Can you add a header stating the license (Apache) and authors?

**Pull request merged**